### PR TITLE
Closes #2255: Adds onBackPressListener to editToolbar

### LINKED ISSUE #2255: Implement better handling of back press when keyboard is showing on search bar
Per recommendation [here](https://github.com/mozilla-mobile/fenix/issues/847), it'd be great if our "hide keyboard" button also called `onBackPressed`.

It seems like we can achieve this by [listening to `KeyEvent.BACK`](https://stackoverflow.com/questions/8080711/android-back-key-event-while-displaying-soft-keyboard)

### PR DESCRIPTION
**Pull request by sblatz**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### CODE REVIEW

File: components/browser/toolbar/src/test/java/mozilla/components/browser/toolbar/edit/EditToolbarTest.kt [hunk 1]
```diff
@@ -84,6 +84,29 @@ class EditToolbarTest {
...
+        val listener = { listenerInvoked = true }
+
+        val toolbar = BrowserToolbar(context)
+        toolbar.editToolbar.urlView.onAttachedToWindow()
+        toolbar.setOnBackPressListener(listener)
+        toolbar.editToolbar.urlView.setText("https://www.mozilla.org")
+
+        assertEquals(listener, toolbar.editToolbar.onBackPressedListener)
+
+        // TODO this crashes as the shadow library does not have KeyEvent.KEYCODE_BACK
```
Comment (sblatz) [inline components/browser/toolbar/src/test/java/mozilla/components/browser/toolbar/edit/EditToolbarTest.kt]: The Shadow version of KeyEvents does not seem to support KEYCODE_BACK. Is there any way around this?

image

**Comment by sblatz** [comment]
Made the changes requested and ready for review!

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/edit/EditToolbar.kt [hunk 2]
```diff
@@ -61,6 +62,13 @@ class EditToolbar(
             updateClearViewVisibility(text)
             editListener?.onTextChanged(text)
         }
+
+        setOnDispatchKeyEventPreImeListener { event ->
+            if (event?.keyCode == KeyEvent.KEYCODE_BACK) {
+                editListener?.onCancelEditing()
```
Comment (pocmo) [inline components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/edit/EditToolbar.kt]: This calls the `editListener` instance directly and `onEditCancelled` in `BrowserToolar` never gets called?

Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/edit/EditToolbar.kt]: Fixed, good catch! My previous test I had written would have caught this, but unfortunately there's no way for us to simulate a `KEYCODE_BACK` press 😕

**Comment by sblatz** [comment]
I believe Taskcluster is failing due to something outside of my control. The error it's displaying is [found here](https://taskcluster-artifacts.net/B_N7NHvTQ1-fjdgFRk5RyQ/0/public/logs/live_backing.log)

image

@pocmo any thoughts?

**Comment by pocmo** [comment]
Looks like it passed now? :)

**Comment by Amejia481** [comment]
@sblatz Looks like an network error, they tend to happen from time to time. In case that happens again you could re-trigger the build by updating the title of the PR, you can add a dot "." add the end of the title.

File: components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt [hunk 3]
```diff
@@ -468,6 +468,12 @@ class BrowserToolbar @JvmOverloads constructor(
         }
     }
 
+    internal fun onEditCancelled() {
+        if (editToolbar.editListener?.onCancelEditing() == true) {
+            displayMode()
+        }
+    }
```
Comment (pocmo) [inline components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt]: Okay, really the last nit... ;)

Should this be:

`if (editToolbar.editListener?.onCancelEditing() != false) {`

Reasoning: We want to switch to display mode if the listener returns `true` OR if there's no listener. I guess now that would only happen if there's a listener and it returns `true`?

Comment (sblatz) [reply components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt]: Fixed :) I was wondering why the one above was doing `!= false` as that just seemed more confusing for no reason. Thanks for the clarification!

File: docs/changelog.md [hunk 4]
```diff
@@ -36,6 +36,11 @@ permalink: /changelog/
 * **concept-fetch**
   * ⚠️ **This is a breaking API change**: the [`Response`](https://mozac.org/api/mozilla.components.concept.fetch/-response/) properties `.success` and `.clientError` were renamed to `.isSuccess` and `isClientError` respectively to match Java conventions.
 
+* **browser-toolbar**
+  * Adds `onCancelEditing` to `onEditListener` in `BrowserToolbar` which is fired when a back button press occurs while the keyboard is displayed.
+    This is especially useful if you want to call `activity.onBackPressed()` to navigate away rather than just dismiss the keyboard.
+    Its return value is used to determine if `displayMode` will switch from edit to view.
+
```
Comment (pocmo) [inline docs/changelog.md]: Oh, damn. One last thing: This needs to be moved up to 0.47.0 now. And then let's merge :)

**Pull request merged**